File: 0345-reverse-vowels-of-a-string/0345-reverse-vowels-of-a-string.java
import java.util.*;

class Solution {
    public String reverseVowels(String s) {
        // IceCreAm
        //    ^^
        // search towards middle. if found vowel, stop and swap lt ++ rt -- 
        // until crossed
        // no more vowel -> have to check each iteration if lt and rt crossed
        int lt = 0;
        int rt = s.length() - 1;
        char[] chars = s.toCharArray();

        while (lt < rt) {
            // move lt
            while (!isVowel(chars[lt])) {
                lt++;

                if (lt >= rt) {
                    break;
                }
            }
            // move rt
            while (!isVowel(chars[rt])) {
                rt--;

                if (lt >= rt) {
                    break;
                }
            }
            if (lt >= rt) {
                break;
            }
            // swap
            swap(chars, lt, rt);
            lt++;
            rt--;
        }
        return new String(chars);
    }

    private void swap(char[] chars, int a, int b) {
        char tmp;
        tmp = chars[a];
        chars[a] = chars[b];
        chars[b] = tmp;
    }

    private boolean isVowel(char c) {
        //TODO make a list and use contains.
        if (c == 'a' || c == 'e' || c == 'i' || c == 'o' || c == 'u'
            || c == 'A' || c == 'E' || c == 'I' || c == 'O' || c == 'U') {
            return true;
        }
        return false;
    }
}

File: 0345-reverse-vowels-of-a-string/0345-reverse-vowels-of-a-string.py
class Solution:
    def reverseVowels(self, s: str) -> str:
        vowels = set("aeiouAEIOU")
        chars = list(s)

        lt, rt = 0 , len(s) - 1

        while lt < rt:
            if chars[lt] not in vowels:
                lt += 1
                continue
            if chars[rt] not in vowels:
                rt -= 1
                continue

            chars[lt], chars[rt] = chars[rt], chars[lt]
            lt +=1 
            rt -=1
        
        return ''.join(chars)